File: app/src/main/java/com/dpis/module/appconfig/EditorActions.java
package com.dpis.module.appconfig;

import com.dpis.module.appconfig.EditorDraft;
import com.dpis.module.appconfig.EditorPresentation;
import com.dpis.module.ConfigEditorDestination;
import com.dpis.module.applist.AppListItem;

/**
 * Builds the per-app editor actions without owning Activity state or side effects.
 *
 * The host is deliberately capability-shaped: Activities keep ownership of persistence, dialogs,
 * runtime work, and refresh timing while this controller owns the immutable draft transitions.
 */
public final class EditorActions {
    public interface Host {
        void updateDraft(EditorDraft draft);
        void showWechatDpiHelp();
        void navigate(ConfigEditorDestination destination);
        void toggleScope(boolean currentlySelected, Runnable onSelected, Runnable onDeselected);
        boolean setDpisEnabled(boolean enabled);
        void executeProcessAction(AppConfigDialogBinder.ProcessAction action);
        void startFeedbackDiagnostic(EditorDraft draft);
        void save(EditorDraft draft);
        void close();
    }

    private EditorActions() {
    }

    public static EditorPresentation.Actions create(
            Host host,
            AppListItem item,
            EditorDraft draft
    ) {
        return new EditorPresentation.Actions() {
            @Override public void updateViewportInput(String value) {
                host.updateDraft(draft.withViewportInput(draft.viewportMode, value));
            }

            @Override public void changeViewportMode(String targetType) {
                host.updateDraft(draft.withViewportMode(targetType));
            }

            @Override public void updateFontInput(String value) {
                host.updateDraft(draft.withFontInput(value));
            }

            @Override public void changeFontMode(String mode) {
                host.updateDraft(draft.withFontMode(mode));
            }

            @Override public void updateWechatDpiInput(String value) {
                host.updateDraft(draft.withWechatDpiInput(value));
            }

            @Override public void showWechatDpiHelp() {
                host.showWechatDpiHelp();
            }

            @Override public void updateTypeface(String typefaceId) {
                host.updateDraft(draft.withAdvancedConfig(
                        typefaceId,
                        draft.draftFontHookDomainsRaw,
                        draft.viewportApplyMode,
                        draft.fontHookDomainsResetRequested,
                        draft.viewportApplyModeResetRequested));
            }

            @Override public void updateHookChain(
                    String rawDomains,
                    boolean resetDomains,
                    String viewportApplyMode,
                    boolean resetViewportApplyMode
            ) {
                host.updateDraft(draft.withAdvancedConfig(
                        draft.selectedTypefaceId,
                        rawDomains,
                        viewportApplyMode,
                        resetDomains,
                        resetViewportApplyMode));
            }

            @Override public void navigate(ConfigEditorDestination destination) {
                host.navigate(destination);
            }

            @Override public void reset() {
                host.updateDraft(draft.cleared());
            }

            @Override public void toggleScope() {
                host.toggleScope(
                        draft.scopeSelected,
                        () -> host.updateDraft(draft.withScopeSelected(true)),
                        () -> host.updateDraft(draft.withScopeSelected(false)));
            }

            @Override public void toggleDpisEnabled() {
                boolean enabled = !draft.dpisEnabled;
                if (host.setDpisEnabled(enabled)) {
                    host.updateDraft(draft.withDpisEnabled(enabled));
                }
            }

            @Override public void startProcess() {
                host.executeProcessAction(AppConfigDialogBinder.ProcessAction.START);
            }

            @Override public void restartProcess() {
                host.executeProcessAction(AppConfigDialogBinder.ProcessAction.RESTART);
            }

            @Override public void stopProcess() {
                host.executeProcessAction(AppConfigDialogBinder.ProcessAction.STOP);
            }

            @Override public void startFeedbackDiagnostic() {
                host.startFeedbackDiagnostic(draft);
            }

            @Override public void save() {
                host.save(draft);
            }

            @Override public void close() {
                host.close();
            }
        };
    }
}

File: app/src/main/java/com/dpis/module/appconfig/editor/EditorActions.kt
package com.dpis.module.appconfig.editor

import com.dpis.module.appconfig.AppConfigProcessAction
import com.dpis.module.applist.AppListItem
import com.dpis.module.ui.ConfigEditorDestination

/** Builds immutable-draft actions without owning Activity state or side effects. */
object EditorActions {
    interface Host {
        fun updateDraft(draft: EditorDraft)
        fun resetDraft()
        fun showWechatDpiHelp()
        fun navigate(destination: ConfigEditorDestination)
        fun toggleScope(currentlySelected: Boolean, onSelected: Runnable, onDeselected: Runnable)
        fun setDpisEnabled(enabled: Boolean): Boolean
        fun executeProcessAction(action: AppConfigProcessAction)
        fun startFeedbackDiagnostic(draft: EditorDraft)
        fun save(draft: EditorDraft)
        fun close()
    }

    @JvmStatic
    fun create(host: Host, item: AppListItem, draft: EditorDraft): EditorPresentation.Actions =
        object : EditorPresentation.Actions {
            override fun updateViewportInput(value: String) =
                host.updateDraft(draft.withViewportInput(draft.viewportMode, value))

            override fun changeViewportMode(targetType: String) =
                host.updateDraft(draft.withViewportMode(targetType))

            override fun updateFontInput(value: String) = host.updateDraft(draft.withFontInput(value))
            override fun changeFontMode(mode: String) = host.updateDraft(draft.withFontMode(mode))
            override fun updateWechatDpiInput(value: String) =
                host.updateDraft(draft.withWechatDpiInput(value))

            override fun showWechatDpiHelp() = host.showWechatDpiHelp()
            override fun updateTypeface(typefaceId: String) = host.updateDraft(draft.withAdvancedConfig(
                typefaceId, draft.draftFontHookDomainsRaw, draft.viewportApplyMode,
                draft.fontHookDomainsResetRequested, draft.viewportApplyModeResetRequested,
            ))

            override fun updateHookChain(
                rawDomains: String,
                resetDomains: Boolean,
                viewportApplyMode: String,
                resetViewportApplyMode: Boolean,
            ) = host.updateDraft(draft.withAdvancedConfig(
                draft.selectedTypefaceId, rawDomains, viewportApplyMode,
                resetDomains, resetViewportApplyMode,
            ))

            override fun navigate(destination: ConfigEditorDestination) = host.navigate(destination)
            override fun reset() = host.resetDraft()
            override fun toggleScope() = host.toggleScope(
                draft.scopeSelected,
                Runnable { host.updateDraft(draft.withScopeSelected(true)) },
                Runnable { host.updateDraft(draft.withScopeSelected(false)) },
            )
            override fun toggleDpisEnabled() {
                val enabled = !draft.dpisEnabled
                if (host.setDpisEnabled(enabled)) host.updateDraft(draft.withDpisEnabled(enabled))
            }
            override fun startProcess() = host.executeProcessAction(AppConfigProcessAction.START)
            override fun restartProcess() = host.executeProcessAction(AppConfigProcessAction.RESTART)
            override fun stopProcess() = host.executeProcessAction(AppConfigProcessAction.STOP)
            override fun startFeedbackDiagnostic() = host.startFeedbackDiagnostic(draft)
            override fun save() = host.save(draft)
            override fun close() = host.close()
        }
}
